ddtrace/appsec/_ddwaf.pyx

```diff
@@ -0,0 +1,255 @@
+        cdef ddwaf_object *ptr
+        cdef ddwaf_object *obj
+
+        idx = self._next_idx
+        if idx + n > self._size:
+            self._size += ((1000 - (n % 1000)) % 1000)
+            ptr = <ddwaf_object *> PyMem_Realloc(self._ptr, self._size * sizeof(ddwaf_object))
+            if ptr == NULL:
+                raise MemoryError
+            memset(ptr + idx, 0, (self._size - idx) * sizeof(ddwaf_object))
+            if self._ptr != NULL and ptr != self._ptr:
+                # we need to patch all array objects because they use pointers to other objects
+                for i in range(idx):
+                    obj = ptr + i
+                    if (obj.type == DDWAF_OBJ_TYPE.DDWAF_OBJ_MAP or obj.type == DDWAF_OBJ_TYPE.DDWAF_OBJ_ARRAY) and obj.array != NULL:
+                        obj.array = obj.array - self._ptr + ptr
+            self._ptr = ptr
+        self._next_idx += n
+        return idx
+
+    cdef int _make_string(self, ssize_t idx, object string) except -1:
+        cdef const char * ptr
+        cdef ssize_t length
+        cdef ddwaf_object *obj
+
+        if isinstance(string, six.binary_type):
+            ptr = PyBytes_AsString(string)
+            length = PyBytes_Size(string)
+        elif isinstance(string, six.text_type):
+            IF PY_MAJOR_VERSION >= 3:
+                ptr = PyUnicode_AsUTF8AndSize(string, &length)
+            if ptr == NULL:
+                string = PyUnicode_AsEncodedString(string, "utf-8", "surrogatepass")
+                ptr = PyBytes_AsString(string)
+                length = PyBytes_Size(string)
+        self._string_refs.append(string)
+
+        obj = self._ptr + idx
+        obj.type = DDWAF_OBJ_TYPE.DDWAF_OBJ_STRING
+        obj.stringValue = ptr
+        obj.nbEntries = length
+        return 0
+
+    cdef void _make_array(self, ssize_t idx, ssize_t array_idx, ssize_t nb_entries):
+        cdef ddwaf_object *obj
+        obj = self._ptr + idx
+        obj.type = DDWAF_OBJ_TYPE.DDWAF_OBJ_ARRAY
+        obj.array = self._ptr + array_idx
+        obj.nbEntries = nb_entries
+
+    cdef void _make_map(self, ssize_t idx, ssize_t array_idx, ssize_t nb_entries):
+        cdef ddwaf_object *obj
+        obj = self._ptr + idx
+        obj.type = DDWAF_OBJ_TYPE.DDWAF_OBJ_MAP
+        obj.array = self._ptr + array_idx
+        obj.nbEntries = nb_entries
+
+    cdef int _set_parameter(self, ssize_t idx, object string) except -1:
+        cdef const char * ptr
+        cdef ssize_t length
+        cdef ddwaf_object *obj
+
+        if isinstance(string, six.binary_type):
+            ptr = PyBytes_AsString(string)
+            length = PyBytes_Size(string)
+        else:
+            IF PY_MAJOR_VERSION >= 3:
+                ptr = PyUnicode_AsUTF8AndSize(string, &length)
+            if ptr == NULL:
+                string = PyUnicode_AsEncodedString(string, "utf-8", "surrogatepass")
+                ptr = PyBytes_AsString(string)
+                length = PyBytes_Size(string)
```

Comment:
This code is duplicated, can you at least create a `cdef inline` for this?

Reply:
@brettlangdon I tried a few stuff but did not manage :/ Can we keep it as is for the time being and I'll grab someone to help me there after? :)